File: src/main/scala/parameters.scala
package faba.parameters

import org.objectweb.asm.Type
import org.objectweb.asm.tree.analysis.{BasicInterpreter, Frame, BasicValue}
import org.objectweb.asm.tree.{MethodInsnNode, TypeInsnNode, JumpInsnNode, AbstractInsnNode}
import org.objectweb.asm.Opcodes._

import faba.analysis.{Utils => AnalysisUtils, _}
import faba.cfg._
import faba.data._
import faba.engine._

object `package` {

  type SoP = Set[Set[Key]]

  implicit class SopOps(val sop1: SoP) {
    def join(sop2: SoP): SoP =
      sop1 ++ sop2

    def meet(sop2: SoP): SoP =
      for {
        prod1 <- sop1
        prod2 <- sop2
      } yield prod1 ++ prod2
  }

}

sealed trait Result
case object Identity extends Result
case object Error extends Result
case object Return extends Result
case object NPE extends Result
case class ConditionalNPE(sop: SoP) extends Result {
  if (sop.map(_.size).sum > 30) throw new LimitReachedException
}

object ConditionalNPE {
  def apply(passing: Key): ConditionalNPE = ConditionalNPE(Set(Set(passing)))
}

object Result {

  def join(r1: Result, r2: Result): Result = (r1, r2) match {
    case (Error, _) => r2
    case (_, Error) => r1
    case (Identity, _) => r2
    case (_, Identity) => r1
    case (Return, _) => Return
    case (_, Return) => Return
    case (NPE, NPE) => NPE
    case (NPE, r2: ConditionalNPE) => r2
    case (r1: ConditionalNPE, NPE) => r1
    case (ConditionalNPE(e1), ConditionalNPE(e2)) => ConditionalNPE(e1 join e2)
  }

  def combineNullable(r1: Result, r2: Result): Result = (r1, r2) match {
    case (Error, _) => r2
    case (_, Error) => r1
    case (Identity, _) => r2
    case (_, Identity) => r1
    case (Return, _) => r2
    case (_, Return) => r1
    case (NPE, _) => NPE
    case (_, NPE) => NPE
    case (ConditionalNPE(e1), ConditionalNPE(e2)) => ConditionalNPE(e1 join e2)
  }

  def meet(r1: Result, r2: Result): Result = (r1, r2) match {
    case (Identity, _) => r2
    case (Return, _) => r2
    case (_, Return) => r1
    case (NPE, _) => NPE
    case (_, NPE) => NPE
    case (Error, _) => Error
    case (_, Error) => Error
    // Oops - cycle dominates
    case (_, Identity) => Identity
    case (ConditionalNPE(e1), ConditionalNPE(e2)) => ConditionalNPE(e1 meet e2)
  }

  def subMeet(r1: Result, r2: Result): Result = (r1, r2) match {
    case (Identity, _) => r2
    case (_, Identity) => r1
    case (Return, _) => r2
    case (_, Return) => r1
    case (NPE, _) => NPE
    case (_, NPE) => NPE
    case (Error, _) => Error
    case (_, Error) => Error
    case (ConditionalNPE(e1), ConditionalNPE(e2)) => ConditionalNPE(e1 meet e2)
  }
}

object NotNullInConstraint {
  // (taken: Boolean, hasCompanions: Boolean)
  val TAKEN = 1 // 1 << 0
  val HAS_COMPANIONS = 2 //1 << 1

  // (x != null) / (x instanceOf ..) was proceeded
  def isTaken(constraint: Int) = (constraint & TAKEN) != 0

  // some previous configuration is foo(param) -
  // hasCompanions = true means that it is possible indirect dereference
  def hasCompanions(constraint: Int) = (constraint & HAS_COMPANIONS) != 0
}

object NotNullInAnalysis {
  // For inference of @NotNull parameters we need to combine information from all configurations,
  // not just from leaf configurations.
  // So we need to build and traverse graph of configurations in some tricky order.
  // For this purpose we maintain "pending list" of what to do.

  /**
   * Element of pending list - pending action.
   */
  sealed trait PendingAction

  /**
   * "Forward step". Driving of current configuration.
   * @param state state to process
   *
   * @see faba.parameters.NotNullInAnalysis#processState(faba.analysis.State)
   */
  case class ProceedState(state: State) extends PendingAction

  /**
   * "Combine step". Calculates "approximation" (see the paper) of behavior from some configuration.
   * Given local delta and sub-sigmas, computes current sigma.
   *
   * A note about states:
   * there is a micro optimization: driving of straight section/piece of graph of configurations without push/pop of actions.
   * States are states/configurations from this straight section of a graph.
   *
   * @param states a list of states representing the straight section of a graph.
   * @param subResult - local delta (see the paper), produced by driving of straight section.
   * @param indices - indices of child states
   */
  case class MakeResult(states: List[State], subResult: Result, indices: List[Int]) extends PendingAction

  /**
   * Reusable pending list (pending stack) for push/pop actions during analyses.
   * @see faba.parameters.NotNullInAnalysis#pending
   */
  val sharedPendingStack = new Array[PendingAction](LimitReachedException.limit)

  /**
   * Reusable storage of sub results during analyses.
   * @see faba.parameters.NotNullInAnalysis#results
   */
  val sharedResults = new Array[Result](LimitReachedException.limit)
}

class NotNullInAnalysis(val richControlFlow: RichControlFlow, val direction: Direction, val stable: Boolean) extends Analysis[Result] {
  import NotNullInAnalysis._

  val results = NotNullInAnalysis.sharedResults
  val pending = NotNullInAnalysis.sharedPendingStack

  /**
   *
   * @param delta
   * @param subResults
   * @return
   */
  def combineResults(delta: Result, subResults: List[Result]): Result =
    Result.meet(delta, subResults.reduce(Result.join))

  override def mkEquation(result: Result): Equation[Key, Value] = result match {
    case Identity | Return | Error =>
      Equation(aKey, Final(Values.Top))
    case NPE =>
      Equation(aKey, Final(Values.NotNull))
    case ConditionalNPE(cnf) =>
      Equation(aKey, Pending(cnf.map(p => Component(Values.Top, p))))
  }

  var npe = false

  private var pendingTop: Int = 0

  final def pendingPush(action: PendingAction) {
    pending(pendingTop) = action
    pendingTop += 1
  }

  final def pendingPop(): PendingAction = {
    pendingTop -= 1
    pending(pendingTop)
  }

  def analyze(): Equation[Key, Value] = {
    pendingPush(ProceedState(createStartState()))

    while (pendingTop > 0 && earlyResult.isEmpty) pendingPop() match {
      case MakeResult(states, delta, subIndices) =>
        val result = combineResults(delta, subIndices.map(results))
        for (state <- states) {
          val insnIndex = state.conf.insnIndex
          results(state.index) = result
          computed(insnIndex) = state :: computed(insnIndex)
        }
      case ProceedState(state) =>
        processState(state)
    }

    mkEquation(earlyResult.getOrElse(results(0)))
  }

  override def processState(fState: State): Unit = {

    var state = fState
    var states: List[State] = Nil
    var subResult: Result = Identity

    while (true) {
      computed(state.conf.insnIndex).find(prevState => AnalysisUtils.stateEquiv(state, prevState)) match {
        case Some(ps) =>
          results(state.index) = results(ps.index)
          if (states.nonEmpty)
            pendingPush(MakeResult(states, subResult, List(ps.index)))
          return
        case None =>
      }


      val stateIndex = state.index
      val conf = state.conf
      val insnIndex = conf.insnIndex
      val history = state.history

      val fold = dfsTree.loopEnters(insnIndex) && history.exists(prevConf => AnalysisUtils.isInstance(conf, prevConf))

      if (fold) {
        results(stateIndex) = Identity
        computed(insnIndex) = state :: computed(insnIndex)
        if (states.nonEmpty)
          pendingPush(MakeResult(states, subResult, List(stateIndex)))
        return
      }

      val frame = conf.frame
      val insnNode = methodNode.instructions.get(insnIndex)
      val nextHistory = if (dfsTree.loopEnters(insnIndex)) conf :: history else history
      val (nextFrame, localSubResult) = execute(frame, insnNode)
      val notEmptySubResult = localSubResult != Identity

      // local "summing"
      val subResult2 = Result.subMeet(subResult, localSubResult)
      val noSwitch = subResult == subResult2
      subResult = subResult2

      if (localSubResult == NPE) {
        // npe was detected, storing this fact for further analyses
        npe = true
        results(stateIndex) = NPE
        computed(insnIndex) = state :: computed(insnIndex)
        pendingPush(MakeResult(states, subResult, List(stateIndex)))
        return
      }

      val constraint = state.constraint
      // the constant will be inlined by javac
      val companionsNow = if (localSubResult != Identity) 2 else 0
      val constraint1 = constraint | companionsNow

      insnNode.getOpcode match {
        case ARETURN | IRETURN | LRETURN | FRETURN | DRETURN | RETURN =>
          if (!NotNullInConstraint.hasCompanions(constraint)) {
            earlyResult = Some(Return)
            return
          } else {
            results(stateIndex) = Return
            computed(insnIndex) = state :: computed(insnIndex)
            // important to put subResult
            if (states.nonEmpty)
              pendingPush(MakeResult(states, subResult, List(stateIndex)))
            return
          }
        case ATHROW if NotNullInConstraint.isTaken(constraint) =>
          results(stateIndex) = NPE
          npe = true
          computed(insnIndex) = state :: computed(insnIndex)
          if (states.nonEmpty)
            pendingPush(MakeResult(states, subResult, List(stateIndex)))
          return
        case ATHROW =>
          results(stateIndex) = Error
          computed(insnIndex) = state :: computed(insnIndex)
          if (states.nonEmpty)
            pendingPush(MakeResult(states, subResult, List(stateIndex)))
          return
        case IFNONNULL if popValue(frame).isInstanceOf[ParamValue] =>
          val nextInsnIndex = insnIndex + 1
          val nextState = State(genId(), Conf(nextInsnIndex, nextFrame), nextHistory, constraint1 | 1)
          states = state :: states
          state = nextState
        case IFNULL if popValue(frame).isInstanceOf[ParamValue] =>
          val nextInsnIndex = methodNode.instructions.indexOf(insnNode.asInstanceOf[JumpInsnNode].label)
          val nextState = State(genId(), Conf(nextInsnIndex, nextFrame), nextHistory, constraint1 | 1)
          states = state :: states
          state = nextState
        case IFEQ if popValue(frame).isInstanceOf[ParamInstanceOfCheckValue] =>
          val nextInsnIndex = methodNode.instructions.indexOf(insnNode.asInstanceOf[JumpInsnNode].label)
          val nextState = State(genId(), Conf(nextInsnIndex, nextFrame), nextHistory, constraint1 | 1)
          states = state :: states
          state = nextState
        case IFNE if popValue(frame).isInstanceOf[ParamInstanceOfCheckValue] =>
          val nextInsnIndex = insnIndex + 1
          val nextState = State(genId(), Conf(nextInsnIndex, nextFrame), nextHistory, constraint1 | 1)
          states = state :: states
          state = nextState
        case _ =>
          val nextInsnIndices = controlFlow.transitions(insnIndex)
          val nextStates = nextInsnIndices.map {
            nextInsnIndex =>
              val nextFrame1 = if (controlFlow.errors(nextInsnIndex) && controlFlow.errorTransitions(insnIndex -> nextInsnIndex)) {
                val handler = new Frame(frame)
                handler.clearStack()
                handler.push(new BasicValue(Type.getType("java/lang/Throwable")))
                handler
              } else {
                nextFrame
              }
              State(genId(), Conf(nextInsnIndex, nextFrame1), nextHistory, constraint1)
          }
          states = state :: states
          if (nextStates.size == 1 && noSwitch) {
            state = nextStates.head
          } else {
            pendingPush(MakeResult(states, subResult, nextStates.map(_.index)))
            nextStates.foreach {s => pendingPush(ProceedState(s))}
            return
          }
      }
    }
  }

  private def execute(frame: Frame[BasicValue], insnNode: AbstractInsnNode) = insnNode.getType match {
    case AbstractInsnNode.LABEL | AbstractInsnNode.LINE | AbstractInsnNode.FRAME =>
      (frame, Identity)
    case _ =>
      val nextFrame = new Frame(frame)
      NonNullInterpreter.reset()
      nextFrame.execute(insnNode, NonNullInterpreter)
      (nextFrame, NonNullInterpreter.getSubResult)
  }
}

case class NullableInConstraint(taken: Boolean)

object NullableInConstraint {
  val TAKEN = 1
  val NOT_TAKEN = 0
}

object NullableInAnalysis {
  val sharedPendingStack = new Array[State](LimitReachedException.limit)
}

class NullableInAnalysis(val richControlFlow: RichControlFlow, val direction: Direction, val stable: Boolean) extends Analysis[Result] {

  val pending = NullableInAnalysis.sharedPendingStack

  override def mkEquation(result: Result): Equation[Key, Value] = result match {
    case NPE => Equation(aKey, Final(Values.Top))
    case Identity | Return | Error => Equation(aKey, Final(Values.Null))
    case ConditionalNPE(cnf) =>
      Equation(aKey, Pending(cnf.map(p => Component(Values.Top, p))))
  }

  private var myResult: Result = Identity

  private var pendingTop: Int = 0

  final def pendingPush(state: State) {
    pending(pendingTop) = state
    pendingTop += 1
  }

  final def pendingPop(): State = {
    pendingTop -= 1
    pending(pendingTop)
  }

  def analyze(): Equation[Key, Value] = {
    pendingPush(createStartState())

    while (pendingTop > 0 && earlyResult.isEmpty)
      processState(pendingPop())

    mkEquation(earlyResult.getOrElse(myResult))
  }

  override def processState(fState: State): Unit = {

    var state = fState

    while (true) {
      computed(state.conf.insnIndex).find(prevState => AnalysisUtils.stateEquiv(state, prevState)) match {
        case Some(ps) =>
          return
        case None =>
      }

      val conf = state.conf
      val insnIndex = conf.insnIndex
      val history = state.history

      val isLoopEnter = dfsTree.loopEnters(insnIndex)
      val fold = isLoopEnter && history.exists(prevConf => AnalysisUtils.isInstance(conf, prevConf))

      computed(insnIndex) = state :: computed(insnIndex)

      if (fold)
        return

      val taken = state.constraint
      val frame = conf.frame
      val insnNode = methodNode.instructions.get(insnIndex)
      val nextHistory = if (isLoopEnter) conf :: history else history
      val (nextFrame, localSubResult) = execute(frame, insnNode)

      if (localSubResult == NPE) {
        earlyResult = Some(NPE)
        return
      }

      if (localSubResult.isInstanceOf[ConditionalNPE])
        myResult = Result.combineNullable(myResult, localSubResult)

      insnNode.getOpcode match {
        case ARETURN | IRETURN | LRETURN | FRETURN | DRETURN | RETURN =>
          if (insnNode.getOpcode == ARETURN && popValue(frame).isInstanceOf[ParamValue]) {
            earlyResult = Some(NPE)
            return
          }
          return
        case ATHROW if state.constraint == 1 =>
          earlyResult = Some(NPE)
          return
        case ATHROW =>
          return
        case IFNONNULL if popValue(frame).isInstanceOf[ParamValue] =>
          val nextInsnIndex = insnIndex + 1
          state = State(genId(), Conf(nextInsnIndex, nextFrame), nextHistory, 1)
        case IFNULL if popValue(frame).isInstanceOf[ParamValue] =>
          val nextInsnIndex = methodNode.instructions.indexOf(insnNode.asInstanceOf[JumpInsnNode].label)
          state = State(genId(), Conf(nextInsnIndex, nextFrame), nextHistory, 1)
        case IFEQ if popValue(frame).isInstanceOf[ParamInstanceOfCheckValue] =>
          val nextInsnIndex = methodNode.instructions.indexOf(insnNode.asInstanceOf[JumpInsnNode].label)
          state = State(genId(), Conf(nextInsnIndex, nextFrame), nextHistory, 1)
        case IFNE if popValue(frame).isInstanceOf[ParamInstanceOfCheckValue] =>
          val nextInsnIndex = insnIndex + 1
          state = State(genId(), Conf(nextInsnIndex, nextFrame), nextHistory, 1)
        case _ =>
          val nextInsnIndices = controlFlow.transitions(insnIndex)
          val nextStates = nextInsnIndices.map {
            nextInsnIndex =>
              val nextFrame1 = if (controlFlow.errors(nextInsnIndex) && controlFlow.errorTransitions(insnIndex -> nextInsnIndex)) {
                val handler = new Frame(frame)
                handler.clearStack()
                handler.push(new BasicValue(Type.getType("java/lang/Throwable")))
                handler
              } else {
                nextFrame
              }
              State(genId(), Conf(nextInsnIndex, nextFrame1), nextHistory, taken)
          }
          if (nextStates.size == 1) {
            state = nextStates.head
          } else {
            nextStates.foreach {s => pendingPush(s)}
            return
          }
      }
    }
  }

  private def execute(frame: Frame[BasicValue], insnNode: AbstractInsnNode) = insnNode.getType match {
    case AbstractInsnNode.LABEL | AbstractInsnNode.LINE | AbstractInsnNode.FRAME =>
      (frame, Identity)
    case _ =>
      val nextFrame = new Frame(frame)
      NullableInterpreter.reset()
      nextFrame.execute(insnNode, NullableInterpreter)
      (nextFrame, NullableInterpreter.getSubResult)
  }
}

abstract class Interpreter extends BasicInterpreter {
  val nullable: Boolean
  protected var _subResult: Result = Identity
  final def reset(): Unit = {
    _subResult = Identity
  }

  def combine(res1: Result, res2: Result): Result

  def getSubResult: Result =
    _subResult

  final override def unaryOperation(insn: AbstractInsnNode, value: BasicValue): BasicValue = {
    insn.getOpcode match {
      case GETFIELD | ARRAYLENGTH | MONITORENTER if value.isInstanceOf[ParamValue] =>
        _subResult = NPE
      case CHECKCAST if value.isInstanceOf[ParamValue] =>
        return new ParamValue(Type.getObjectType(insn.asInstanceOf[TypeInsnNode].desc))
      case INSTANCEOF if value.isInstanceOf[ParamValue] =>
        return ParamInstanceOfCheckValue()
      case _ =>
    }
    super.unaryOperation(insn, value)
  }

  final override def binaryOperation(insn: AbstractInsnNode, v1: BasicValue, v2: BasicValue): BasicValue = {
    insn.getOpcode match {
      case PUTFIELD
        if v1.isInstanceOf[ParamValue] || (nullable && v2.isInstanceOf[ParamValue]) =>
        _subResult = NPE
      case IALOAD | LALOAD | FALOAD | DALOAD | AALOAD | BALOAD | CALOAD | SALOAD if v1.isInstanceOf[ParamValue] =>
        _subResult = NPE
      case _ =>
    }
    super.binaryOperation(insn, v1, v2)
  }

  final override def ternaryOperation(insn: AbstractInsnNode, v1: BasicValue, v2: BasicValue, v3: BasicValue): BasicValue = {
    insn.getOpcode match {
      case IASTORE | LASTORE | FASTORE | DASTORE | BASTORE | CASTORE | SASTORE
        if v1.isInstanceOf[ParamValue] =>
        _subResult = NPE
      case AASTORE
        if v1.isInstanceOf[ParamValue] || (nullable && v3.isInstanceOf[ParamValue]) =>
        _subResult = NPE
      case _ =>
    }
    super.ternaryOperation(insn, v1, v2, v3)
  }

  final override def naryOperation(insn: AbstractInsnNode, values: java.util.List[_ <: BasicValue]): BasicValue = {
    val opCode = insn.getOpcode
    val static = opCode == INVOKESTATIC
    val shift = if (static) 0 else 1
    if ((opCode == INVOKESPECIAL || opCode == INVOKEINTERFACE || opCode == INVOKEVIRTUAL) && values.get(0).isInstanceOf[ParamValue]) {
      _subResult = NPE
    }
    if (nullable && opCode == INVOKEINTERFACE) {
      for (i <- shift until values.size()) {
        if (values.get(i).isInstanceOf[ParamValue]) {
          _subResult = NPE
          return super.naryOperation(insn, values)
        }
      }
      return super.naryOperation(insn, values)
    }
    opCode match {
      case INVOKESTATIC | INVOKESPECIAL | INVOKEVIRTUAL =>
        val stable = opCode == INVOKESTATIC || opCode == INVOKESPECIAL
        val mNode = insn.asInstanceOf[MethodInsnNode]
        for (i <- shift until values.size()) {
          if (values.get(i).isInstanceOf[ParamValue]) {
            val method = Method(mNode.owner, mNode.name, mNode.desc)
            _subResult = combine(_subResult, ConditionalNPE(Key(method, In(i - shift), stable)))
          }
        }
      case _ =>
    }
    super.naryOperation(insn, values)
  }
}

object NonNullInterpreter extends Interpreter {
  override val nullable = false
  override def combine(res1: Result, res2: Result): Result = Result.meet(res1, res2)

}

object NullableInterpreter extends Interpreter {
  override val nullable = true
  override def combine(res1: Result, res2: Result): Result = Result.join(res1, res2)
}

File: src/main/java/faba/java/Parameters.java
package faba.java;

import org.objectweb.asm.Type;
import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.JumpInsnNode;
import org.objectweb.asm.tree.MethodInsnNode;
import org.objectweb.asm.tree.TypeInsnNode;
import org.objectweb.asm.tree.analysis.AnalyzerException;
import org.objectweb.asm.tree.analysis.BasicInterpreter;
import org.objectweb.asm.tree.analysis.BasicValue;
import org.objectweb.asm.tree.analysis.Frame;

import java.util.*;

import static faba.java.AbstractValues.ParamValue;
import static faba.java.AbstractValues.instanceOfCheckValue;
import static faba.java.Parameters.*;
import static org.objectweb.asm.Opcodes.*;

abstract class Parameters {
    // SoP = sum of products
    static Set<Set<Key>> join(Set<Set<Key>> sop1, Set<Set<Key>> sop2) {
        Set<Set<Key>> sop = new HashSet<Set<Key>>();
        sop.addAll(sop1);
        sop.addAll(sop2);
        return sop;
    }

    static Set<Set<Key>> meet(Set<Set<Key>> sop1, Set<Set<Key>> sop2) {
        Set<Set<Key>> sop = new HashSet<Set<Key>>();
        for (Set<Key> prod1 : sop1) {
            for (Set<Key> prod2 : sop2) {
                Set<Key> prod = new HashSet<>();
                prod.addAll(prod1);
                prod.addAll(prod2);
                sop.add(prod);
            }
        }
        return sop;
    }

    // Results
    interface PResult {}
    static final PResult Identity = new PResult() {
        @Override
        public String toString() {
            return "Identity";
        }
    };
    static final PResult Return = new PResult() {
        @Override
        public String toString() {
            return "Return";
        }
    };
    static final PResult NPE = new PResult() {
        @Override
        public String toString() {
            return "NPE";
        }
    };
    static final class ConditionalNPE implements PResult {
        final Set<Set<Key>> sop;
        public ConditionalNPE(Set<Set<Key>> sop) {
            this.sop = sop;
        }

        public ConditionalNPE(Key key) {
            sop = new HashSet<Set<Key>>();
            Set<Key> prod = new HashSet<Key>();
            prod.add(key);
            sop.add(prod);
        }
    }

    static PResult join(PResult r1, PResult r2) {
        if (Identity == r1) return r2;
        if (Identity == r2) return r1;
        if (Return == r1) return Return;
        if (Return == r2) return Return;
        if (NPE == r1) return r2;
        if (NPE == r2) return r1;
        ConditionalNPE cnpe1 = (ConditionalNPE) r1;
        ConditionalNPE cnpe2 = (ConditionalNPE) r2;
        return new ConditionalNPE(join(cnpe1.sop, cnpe2.sop));
    }

    static PResult meet(PResult r1, PResult r2) {
        if (Identity == r1) return r2;
        if (Identity == r2) return r1;
        if (Return == r1) return r2;
        if (Return == r2) return r1;
        if (NPE == r1) return NPE;
        if (NPE == r2) return NPE;
        ConditionalNPE cnpe1 = (ConditionalNPE) r1;
        ConditionalNPE cnpe2 = (ConditionalNPE) r2;
        return new ConditionalNPE(meet(cnpe1.sop, cnpe2.sop));
    }

}

class NonNullInAnalysis extends Analysis<PResult> {

    private static final NonNullInInterpreter interpreter = new NonNullInInterpreter();
    private final Key parameter;

    protected NonNullInAnalysis(RichControlFlow richControlFlow, Direction direction) {
        super(richControlFlow, direction);
        parameter = new Key(method, direction);
    }

    @Override
    PResult identity() {
        return Identity;
    }

    @Override
    PResult combineResults(PResult delta, List<PResult> subResults) {
        PResult subResult = Identity;
        for (PResult sr : subResults) {
            subResult = join(subResult, sr);
        }
        return meet(delta, subResult);
    }

    @Override
    boolean isEarlyResult(PResult result) {
        return false;
    }

    @Override
    Equation<Key, Value> mkEquation(PResult result) {
        Objects.requireNonNull(result);
        if (Identity == result || Return == result) {
            return new Equation<>(parameter, new Final<Key, Value>(Value.Top));
        }
        else if (NPE == result) {
            return new Equation<>(parameter, new Final<Key, Value>(Value.NotNull));
        }
        else {
            ConditionalNPE condNpe = (ConditionalNPE) result;
            Set<Component<Key>> components = new HashSet<>();
            for (Set<Key> prod : condNpe.sop) {
                components.add(new Component<>(false, prod));
            }
            return new Equation<>(parameter, new Pending<>(Value.NotNull, components));
        }
    }

    private int id = 0;
    private Frame<BasicValue> nextFrame = null;
    private PResult subResult = null;

    @Override
    void processState(State state) throws AnalyzerException {
        int stateIndex = state.index;
        Conf conf = state.conf;
        int insnIndex = conf.insnIndex;
        List<Conf> history = state.history;
        boolean taken = state.taken;
        Frame<BasicValue> frame = conf.frame;
        AbstractInsnNode insnNode = methodNode.instructions.get(insnIndex);
        List<Conf> nextHistory = dfsTree.loopEnters.contains(insnIndex) ? append(history, conf) : history;
        boolean hasCompanions = state.hasCompanions;
        execute(frame, insnNode);

        boolean notEmptySubResult = subResult != Identity;

        if (subResult == NPE) {
            results.put(stateIndex, NPE);
            computed.put(insnIndex, append(computed.get(insnIndex), state));
            return;
        }

        int opcode = insnNode.getOpcode();
        switch (opcode) {
            case ARETURN:
            case IRETURN:
            case LRETURN:
            case FRETURN:
            case DRETURN:
            case RETURN:
                if (!hasCompanions) {
                    earlyResult = Return;
                } else {
                    results.put(stateIndex, Return);
                    computed.put(insnIndex, append(computed.get(insnIndex), state));
                }
                return;
            default:
        }

        if (opcode == ATHROW) {
            if (taken) {
                results.put(stateIndex, NPE);
            } else {
                results.put(stateIndex, Identity);
            }
            computed.put(insnIndex, append(computed.get(insnIndex), state));
            return;
        }

        if (opcode == IFNONNULL ) {
            BasicValue popped = popValue(frame);
            if (popped instanceof ParamValue) {
                int nextInsnIndex = insnIndex + 1;
                State nextState = new State(++id, new Conf(nextInsnIndex, nextFrame), nextHistory, true, hasCompanions || notEmptySubResult);
                pending.push(new MakeResult<PResult>(state, subResult, Collections.singletonList(nextState.index)));
                pending.push(new ProceedState<PResult>(nextState));
                return;
            }
        }

        if (opcode == IFNULL && popValue(frame) instanceof ParamValue) {
            int nextInsnIndex = methodNode.instructions.indexOf(((JumpInsnNode)insnNode).label);
            State nextState = new State(++id, new Conf(nextInsnIndex, nextFrame), nextHistory, true, hasCompanions || notEmptySubResult);
            pending.push(new MakeResult<PResult>(state, subResult, Collections.singletonList(nextState.index)));
            pending.push(new ProceedState<PResult>(nextState));
            return;
        }

        if (opcode == IFEQ && popValue(frame) == instanceOfCheckValue) {
            int nextInsnIndex = methodNode.instructions.indexOf(((JumpInsnNode)insnNode).label);
            State nextState = new State(++id, new Conf(nextInsnIndex, nextFrame), nextHistory, true, hasCompanions || notEmptySubResult);
            pending.push(new MakeResult<PResult>(state, subResult, Collections.singletonList(nextState.index)));
            pending.push(new ProceedState<PResult>(nextState));
            return;
        }

        if (opcode == IFNE && popValue(frame) == instanceOfCheckValue) {
            int nextInsnIndex = insnIndex + 1;
            State nextState = new State(++id, new Conf(nextInsnIndex, nextFrame), nextHistory, true, hasCompanions || notEmptySubResult);
            pending.push(new MakeResult<PResult>(state, subResult, Collections.singletonList(nextState.index)));
            pending.push(new ProceedState<PResult>(nextState));
            return;
        }

        // general case
        List<Integer> nextInsnIndices = controlFlow.transitions[insnIndex];
        List<State> nextStates = new ArrayList<>();
        List<Integer> subIndices = new ArrayList<>();
        for (int nextInsnIndex : nextInsnIndices) {
            Frame<BasicValue> nextFrame1 = nextFrame;
            if (controlFlow.errorTransitions.contains(new Edge(insnIndex, nextInsnIndex))) {
                nextFrame1 = new Frame<>(frame);
                nextFrame1.clearStack();
                nextFrame1.push(new BasicValue(Type.getType("java/lang/Throwable")));
            }
            nextStates.add(new State(++id, new Conf(nextInsnIndex, nextFrame1), nextHistory, taken, hasCompanions || notEmptySubResult));
            subIndices.add(id);
        }

        pending.push(new MakeResult<PResult>(state, subResult, subIndices));
        for (State nextState : nextStates) {
            pending.push(new ProceedState<PResult>(nextState));
        }

    }

    private <A> List<A> append(List<A> xs, A x) {
        ArrayList<A> result = new ArrayList<A>();
        if (xs != null) {
            result.addAll(xs);
        }
        result.add(x);
        return result;
    }


    private void execute(Frame<BasicValue> frame, AbstractInsnNode insnNode) throws AnalyzerException {
        switch (insnNode.getType()) {
            case AbstractInsnNode.LABEL:
            case AbstractInsnNode.LINE:
            case AbstractInsnNode.FRAME:
                nextFrame = frame;
                subResult = Identity;
                break;
            default:
                nextFrame = new Frame<>(frame);
                interpreter.reset();
                nextFrame.execute(insnNode, interpreter);
                subResult = interpreter.getSubResult();
        }
    }
}

class NonNullInInterpreter extends BasicInterpreter {
    private PResult subResult = Identity;
    public PResult getSubResult() {
        return subResult;
    }
    void reset() {
        subResult = Identity;
    }

    @Override
    public BasicValue unaryOperation(AbstractInsnNode insn, BasicValue value) throws AnalyzerException {
        switch (insn.getOpcode()) {
            case GETFIELD:
            case ARRAYLENGTH:
            case MONITORENTER:
                if (value instanceof ParamValue) {
                    subResult = NPE;
                }
                break;
            case CHECKCAST:
                if (value instanceof ParamValue) {
                    return new ParamValue(Type.getObjectType(((TypeInsnNode)insn).desc));
                }
                break;
            case INSTANCEOF:
                if (value instanceof ParamValue) {
                    return instanceOfCheckValue;
                }
                break;
            default:

        }
        return super.unaryOperation(insn, value);
    }

    @Override
    public BasicValue binaryOperation(AbstractInsnNode insn, BasicValue value1, BasicValue value2) throws AnalyzerException {
        switch (insn.getOpcode()) {
            case IALOAD:
            case LALOAD:
            case FALOAD:
            case DALOAD:
            case AALOAD:
            case BALOAD:
            case CALOAD:
            case SALOAD:
            case PUTFIELD:
                if (value1 instanceof ParamValue) {
                    subResult = NPE;
                }
                break;
            default:
        }
        return super.binaryOperation(insn, value1, value2);
    }

    @Override
    public BasicValue ternaryOperation(AbstractInsnNode insn, BasicValue value1, BasicValue value2, BasicValue value3) throws AnalyzerException {
        switch (insn.getOpcode()) {
            case IASTORE:
            case LASTORE:
            case FASTORE:
            case DASTORE:
            case AASTORE:
            case BASTORE:
            case CASTORE:
            case SASTORE:
                if (value1 instanceof ParamValue) {
                    subResult = NPE;
                }
            default:
        }
        return super.ternaryOperation(insn, value1, value2, value3);
    }

    @Override
    public BasicValue naryOperation(AbstractInsnNode insn, List<? extends BasicValue> values) throws AnalyzerException {
        int opcode = insn.getOpcode();
        boolean isStaticInvoke = opcode == INVOKESTATIC;
        int shift = isStaticInvoke ? 0 : 1;
        if (opcode != MULTIANEWARRAY && !isStaticInvoke && values.get(0) instanceof ParamValue) {
            subResult = NPE;
        }
        switch (opcode) {
            case INVOKESTATIC:
            case INVOKESPECIAL:
                MethodInsnNode methodNode = (MethodInsnNode) insn;
                for (int i = shift; i < values.size(); i++) {
                    if (values.get(i) instanceof ParamValue) {
                        Method method = new Method(methodNode.owner, methodNode.name, methodNode.desc);
                        subResult = meet(subResult, new ConditionalNPE(new Key(method, new In(i - shift))));
                    }
                }
            default:
        }
        return super.naryOperation(insn, values);
    }
}
